### ISSUE TITLE
Is there a way to recreate VoxelInstancer on VoxelTerrain instead of VoxelLodTerrain?

### ISSUE THREAD

**Issue by JoseETeixeira** [Zylann/godot_voxel#277]
I'm building a game based on the script demo (https://github.com/Zylann/voxelgame), my idea is to build stuff like houses and temples as larger meshes and then instantiate them in game. 
It seems pretty straightforward to do with VoxelLodTerrain and VoxelInstancer but I'm lost doing it with cubes. 
So far, my solution was to parse the data of my MagicaVoxel file to get color and position of each voxel. Then, add them manually at voxel library, build a generator to aggregate all the voxels and paste with voxel tools.
This solution has some problems: 
1- I can't for the life of me get a structure to only spawn only once
2- Structures often come with missing parts
3- Collision is not rendered (though this one I can probably get working with some elbow grease)

What would help:
- a way to dynamically create voxels and materials directly from the voxel library (and spawn them in game)
- something similar to VoxelInstancer for VoxelTerrain
- any ideas at all on how to do it properly

**Comment by Zylann** [comment]
So I see multiple topics here and several things that I think are not properly thought of.

> my idea is to build stuff like houses and temples as larger meshes and then instantiate them in game.

Ok, so you mean they won't be voxels, right?

> So far, my solution was to parse the data of my MagicaVoxel file to get color and position of each voxel. Then, add them manually at voxel library,

When you work with MagicaVoxel, voxels are colored cubes. So I don't quite understand why you'd use VoxelLibrary here, when there is `VoxelMesherCube` to accept colors directly.

> build a generator to aggregate all the voxels and paste with voxel tools.

Ok now you want to generate voxels? I thought you wanted to instance meshes? Because `VoxelInstancer` is not really about voxels, it's about meshes.

> I can't for the life of me get a structure to only spawn once

Once, as in a single one in the entire world?

> Structures often come with missing parts

Is that because you place it in the generator? Because generators work per block, and it's very common that when spawning structures they will overlap several blocks, not just fit in one.

> Collision is not rendered (though this one I can probably get working with some elbow grease)

No clue what's up with that. If you generate voxels there should be collision, but I'm still not exactly sure what you have done here.

So far I've been considering two methods for generating larger things in a blocky world:

- Instancing:
`VoxelInstancer` can be ported to `VoxelTerrain`, in which case it will only be able to work at LOD0 (any higher value will never spawn). It shoud only need to wire up the appropriate events internally. It is important to note however, that so far the existing instance generator is not scriptable yet, and was thought to run on smooth meshes. Blocky meshes have a lot of faces oriented 90 degrees so at the very least it needs a cutoff to prevent things from spawning sideways (it has such option). Also, currently it is limited to simple models because its first use case was foliage, trees and rocks, and cannot be told to generate something only once in the entire world. In the future it should be able to spawn scenes. Also in theory voxels can be used as a source of instances, but in any case, there are no voxels involved in what it spawns.

- Structure generation: 
This is an approach coming from Minecraft, and is an approach purely based on voxels (there are no meshes or scenes involved). The idea is to define the things to spawn in voxel form: a tree, a house, a fortress. Then the entire process happens inside the generator: in a given block, take the XZ coordinates, hash them, then based on a seeded chance you decide if a thing should spawn in that chunk. If it does, you paste it (or generate it) within the box it is supposed to occupy. It is possible that box overlaps the borders of the block, so this can be solved by also querying neighbor *positions* (not blocks) and see if they *would* generate such structure. If they do, then pick those that overlap with the current block, and place only that bit. This is made possible under the guarantee that the generator is *deterministic*. Which means blocks do not depend on the existence of other blocks, and they can keep generating independently in threads.
I believe I described this in earlier issues, maybe here https://github.com/Zylann/godot_voxel/issues/163#issuecomment-664944892
And I implemented that technique with the trees in https://github.com/Zylann/voxelgame/blob/aa87dabbf16780c923c59e3dadd078d912e458c8/project/blocky_game/generator/generator.gd#L144

**Comment by JoseETeixeira** [comment]
> Ok, so you mean they won't be voxels, right?

I guess I didn't properly explain what I had in mind and the troubles I found when building. 
I would like them to be voxels, however I would like to auto-create voxels from the vox file when importing, so that each voxel is properly placed in the world with auto-generated collisions and so that each voxel on the large structure can be destroyed.

So far, by using the script you built as an example, I couldn't find an option for coloring a specific voxel without changing materials, so I've been doing things the stupid way and adding one material per block (even if they just differ in color) at the voxel library and trying to place them together by building a generator.

> Once, as in a single one in the entire world? If it is a mesh, why not just place it once up-front and not bother with the generator? If the generator also needs to know somehow, you could generate a "global" map roughly describing your world, and then pass that to the generator whose job will be to produce more detailed voxels from it (which would work as long as that map is fixed for the rest of the game)

Okay so this need some idea explaining. The first large structure I'm trying to place will be a prefab city (made with voxels at Magica Voxel). This city NEEDS to be placed where the player will first spawn and only be placed once, as I intend to build a compass and it would effectively be the north pole. Other cities, temples, and so on would also be built this way, and placed inside VoxelTerrain, however, their location will be based of where the first city is, as I intend to build connecting roads in the future and so on. All of this would be built with voxels, not meshes and placed in VoxelTerrain


By reading issue #246 I managed to get it working by using meshes:

**SCENE NODES**
VoxelTerrain -> GrundLoader(Type Node) -> MeshInstance


`tool
extends Node

const _materials = [
	preload("res://blocky_game/blocks/ancient_temple/ancient_temple_material.tres"),
	preload("res://blocky_game/blocks/ancient_temple/ancient_temple_material.tres"),
]

const voxel_library = preload("res://blocky_game/blocks/voxel_library.tres")
onready var _mesh_instance : MeshInstance = $MeshInstance
var _mesher = VoxelMesherCubes.new()
const _CHANNEL = VoxelBuffer.CHANNEL_TYPE

func _ready():
	_generate()

func _generate():
	var voxels = VoxelBuffer.new()
	voxels.set_channel_depth(VoxelBuffer.CHANNEL_COLOR, VoxelBuffer.DEPTH_8_BIT)
	var palette = VoxelColorPalette.new()
	
	var loader = VoxelVoxLoader.new()
	var err = loader.load_from_file("res://blocky_game/blocks/ancient_temple/ancient_temple.vox", voxels, palette)
	if err != OK:
		push_error(str("Error: ", err))
		return

	_mesher.palette = palette
	_mesher.color_mode = VoxelMesherCubes.COLOR_MESHER_PALETTE
	
	var padded_voxels = VoxelBuffer.new()
	padded_voxels.create(
		voxels.get_size().x + 2, 
		voxels.get_size().y + 2, 
		voxels.get_size().z + 2)
	padded_voxels.set_channel_depth(
		VoxelBuffer.CHANNEL_COLOR, voxels.get_channel_depth(VoxelBuffer.CHANNEL_COLOR))
	padded_voxels.copy_channel_from_area(
		voxels, Vector3(), voxels.get_size(), Vector3(1, 1, 1), VoxelBuffer.CHANNEL_COLOR)

	var mesh = _mesher.build_mesh(padded_voxels, _materials)

	_mesh_instance.mesh = mesh
	
	get_parent().set_material(3,_materials[0])
	get_parent().get_voxel_tool().paste(Vector3(0,20,0),padded_voxels,VoxelBuffer.CHANNEL_COLOR,3)`

**Comment by Zylann** [comment]
> I would like them to be voxels, however I would like to auto-create voxels from the vox file when importing, so that each voxel is properly placed in the world with auto-generated collisions and so that each voxel on the large structure can be destroyed.

Ok so this would not be handled by `VoxelInstancer`.

> So far, by using the script you built as an example, I couldn't find an option for coloring a specific voxel without changing materials

Well yeah because that demo uses an entirely different mesher, so it was built around it, with different way of representing voxel values.

> so I've been doing things the stupid way and adding one material per block (even if they just differ in color) at the voxel library and trying to place them together by building a generator.

Indeed even with `VoxelMesherBlocky` this is far from ideal, there is a way to have different colors without changing materials.

> Okay so this need some idea explaining. The first large structure I'm trying to place will be a prefab city (made with voxels at Magica Voxel). This city NEEDS to be placed where the player will first spawn and only be placed once, as I intend to build a compass and it would effectively be the north pole. Other cities, temples, and so on would also be built this way, and placed inside VoxelTerrain, however, their location will be based of where the first city is, as I intend to build connecting roads in the future and so on. All of this would be built with voxels, not meshes and placed in VoxelTerrain

So what I said applies:
Use the `Structures` approach with a predefined set of data roughly describing the world, and intersect the part of it you need when generating each block. It's a very specific need you have after all.

> By reading issue #246 I managed to get it working by using meshes:

Well if you do this you won't be able to edit them. Not without rebuilding the whole pipeline just for them at least. If you wanted collision then you'd have to use https://docs.godotengine.org/en/stable/classes/class_meshinstance.html#class-meshinstance-method-create-trimesh-collision but I think that's relatively inefficient (despite sounding more "simple").

**Comment by JoseETeixeira** [comment]
I'm gonna try my hand at the structures approach, however, shouldn't the code I did using voxel vox loader work? Or better yet, is there a way to get it to work? It would make things much simpler

**Comment by Zylann** [comment]
You can preload your MagicaVoxel stuff when the game start and store voxels somewhere. So far `VoxLoader` purely exists just so `.vox` files can be decoded. What you do with the data is up to you.

**Comment by JoseETeixeira** [comment]
Thanks! I will get back to you when I get it working. I can help building a tutorial for it if you want, I hope I'm not the only dummy struggling haha

**Comment by JoseETeixeira** [comment]
I am either missing some core components needed to make it work or implemented it in a wrong way. When generating blocks, no cities are found. I used this script as a generator. My VoxelTerrain node is set up as follows:

-Stream : VoxelStreamRegions
-Generator : VoxelGeneratorScript (code below)
-Mesher: VoxelMesherBlocks (the same one as the source code from voxel game)

```
class CityDesc:
	var voxels: VoxelBuffer
	var position: Vector3


class WorldDesc:
	var cities := [] # Contains CityDesc instances
	func _init(): # DOES NOT GET AUTO CALLED
		print("STARTING CITIES")
		var voxels = VoxelBuffer.new()
		voxels.set_channel_depth(VoxelBuffer.CHANNEL_COLOR, VoxelBuffer.DEPTH_8_BIT)
		var palette = VoxelColorPalette.new()
		
		var loader = VoxelVoxLoader.new()
		var err = loader.load_from_file("res://blocky_game/blocks/ancient_temple/ancient_temple.vox", voxels, palette)
		if err != OK:
			push_error(str("Error: ", err))
			return
		var grund = CityDesc.new()
		grund.voxels = voxels;
		grund.position = Vector3(0,0,0)
		cities.append(grund);
	func test_call():# DOES NOT GET  CALLED
		print("STARTING CITIES")
		var voxels = VoxelBuffer.new()
		voxels.set_channel_depth(VoxelBuffer.CHANNEL_COLOR, VoxelBuffer.DEPTH_8_BIT)
		var palette = VoxelColorPalette.new()
		
		var loader = VoxelVoxLoader.new()
		var err = loader.load_from_file("res://blocky_game/blocks/ancient_temple/ancient_temple.vox", voxels, palette)
		if err != OK:
			push_error(str("Error: ", err))
			return
		var grund = CityDesc.new()
		grund.voxels = voxels;
		grund.position = Vector3(0,0,0)
		cities.append(grund);
	# Other stuff...


class Generator extends VoxelGeneratorScript:
	# May be set before the game starts, and never modified afterward
	var _world_desc : WorldDesc

	const AIR_MASK = 0

	func _generate_block(voxels:VoxelBuffer, origin_in_voxels:Vector3,lod:int):
		# Generate stuff: ground with noise, trees with procedural structures etc
		# ...
		_world_desc.test_call() ##DOES NOTHING
		# Place pre-generated cities
		var block_box = AABB(origin_in_voxels, Vector3(16,16,16))
		for city in _world_desc.cities:
			var city_box = AABB(city.position, city.voxels.size())
			if city_box.intersects(block_box):
				voxels.paste(city.voxels, city.position - origin_in_voxels, AIR_MASK)
		# Should be ok with a handful of cities
		# buf of course if you have thousands of them
		# you may want to use a smarter data structure than an array.
		# Could also be optimized out if you know the minimum and maximum height they can be

```

**Comment by Zylann** [comment]
You could test your generator in a simple scene without using the whole pipeline. Just pick a block position that you know is supposed to line up with a city, and check what happens. You can also debug your code this way (since GDScript debugger doesnt support threads).

**Comment by JoseETeixeira** [comment]
I finally managed to do it but still have no materials. I might make a tutorial for begginers using this when I get it to work properly
![2021-06-10_16-06](https://user-images.githubusercontent.com/9901373/121582885-f0e32f80-ca05-11eb-9550-ddbe4c69ec7d.png)

**Comment by Zylann** [comment]
>  I made a cube material in VoxelLibrary and a new empty material on it

`VoxelLibrary` cannot hold materials so I don't know what that means. It does not use materials from meshes either. Materials are configured in an array on `VoxelTerrain` and their index corresponds to the `material_id` property in `Voxel` resources.

> how would I set specific vertex colors (aside from toggling the use vertex color as albedo flag) ?

> How would I a)Set a color for a specific vertex 

MagicaVoxel does not allow you to even do that, so I dont understand why you want this after loading your city. But you can, by making the mesh of the `Voxel` in Blender, vertex paint it, export in a format that can hold vertex colors (dunno if OBJ works, otherwise GLTF) and assign the mesh in `Voxel`. If that doesnt work it could be made working, but I never seen a need for that use case, since a voxel either is a single color, or uses a texture.

> Make things like leaves and things that need transparency work

Once again I have no clue how you expect MagicaVoxel to give you textured voxels such as leaves, but you have an example of this already in the demo. The voxel resource named `leaves` in `VoxelLibrary` has a material ID corresponding to a specific slot on `VoxelTerrain`, in which a specific material was used. That material renders voxels with alpha clip on transparent parts of their texture.
You will also note that there is a third material for semi-transparent voxels such as water.
None of these materials are specific to a given voxel though, because each voxel uses a texture atlas. They are UV-mapped such that they use only a part of a larger texture, so only 3 materials are needed in total.
For texture-less voxels which just have a single color (like ALL of those you get from a `.vox` file), you'd need a material without a texture and `voxel as albedo` enabled, and maybe another for semi-transparent version.

> Would I need to change the generators and stream to get a different mesher working?

Not entirely, depends what you are actually doing. You can change a mesher without touching the generator just fine. And if you use `VoxelMesherCubes` with the same color palette you use in MagicaVoxel, then your city generation should pretty much give you the same output.
The problem is, the existing demo stuff was written for a different strategy. The demo uses voxel types, while you seem to be using colors. You would have to change the logic so that instead of writing types in the `TYPE` channel, it writes corresponding color values into the `COLOR` channel.

**Comment by JoseETeixeira** [comment]
> VoxelLibrary cannot hold materials so I don't know what that means. It does not use materials from meshes either. Materials are configured in an array on VoxelTerrain and their index corresponds to the material_id property in Voxel resources.
Exactly, `VoxelTerrain`, my bad.

The idea was to take the demo as a base model, giving me a full procedural world that doesn't depend only on colors, while my larger instances would be color-based. 

> Once again I have no clue how you expect MagicaVoxel to give you textured voxels such as leaves, but you have an example of this already in the demo. The voxel resource named leaves in VoxelLibrary has a material ID corresponding to a specific slot on VoxelTerrain, in which a specific material was used. That material renders voxels with alpha clip on transparent parts of their texture.

I couldn't get it to work by using `VoxelVoxInstancer` so I ended up downloading Voxenko (https://gaidamakua.itch.io/voxenko), importing my MagicaVoxel model, exporting it as a json with position and colors, then parsing this resulting json, while setting the material to this new material I talked about. 

> Not entirely, depends what you are actually doing. You can change a mesher without touching the generator just fine. And if you use VoxelMesherCubes with the same color palette you use in MagicaVoxel, then your city generation should pretty much give you the same output.
The problem is, the existing demo stuff was written for a different strategy. The demo uses voxel types, while you seem to be using colors. You would have to change the logic so that instead of writing types in the TYPE channel, it writes corresponding color values into the COLOR channel.

From what I understood I would have to either rewrite the demo using only colors and `VoxelMesherCubes` or change the addon code to get colors instead of IDs, is that right?

**Comment by Zylann** [comment]
> From what I understood I would have to either rewrite the demo using only colors and VoxelMesherCubes or change the addon code to get colors instead of IDs, is that right?

If you were to go for a fully color-based terrain, yes.

> can I use both TYPE and COLOR ?

This is currently not possible. It might be added in the future since I see a lot of people trying to bring MagicaVoxel stuff into terrains, but very often I wonder wether or not they realise that Magica and Minecraft voxels have quite different capabilities.
I think one or two devs did something similar by tweaking `VoxelMesherBlocky`, so that on top of mesh batching it would look through other channel to apply coloring from. With Magica it would require to know the palette as well.

Otherwise, for now you'd have to use a `VoxelLibrary` with voxels from the demo, and then a bunch of colored ones whose only purpose is to match colors from MagicaVoxel. Then you'd have to have a mapping between MagicaVoxel values, and these voxel types. In such case, you'd only need to set the `TYPE` channel.

In other words, values you get with `VoxLoader` are in the `COLOR` channel, and correspond to the position on this palette:
![image](https://user-images.githubusercontent.com/1311555/121588753-34519400-ca2e-11eb-8d55-53ff293258d5.png)
So you would need as many `Voxel` resources for every color your city is using.

**Comment by Zylann** [comment]
> Do you know if they publicized it? I would love taking a peek

I don't think they did. But they basically modified this area https://github.com/Zylann/godot_voxel/blob/1d63671680b42d18abcb973c082fd35d41a60e38/meshers/blocky/voxel_mesher_blocky.cpp#L53

> So if I have a 64-color palette I wouldn't be able to add them all, right?

You can have up to 65536 voxel types, so that would actually fit.

> From what I could see the limitations are 8 materials or something like that

Not if you use atlasing and material re-use. You really don't need 64 materials if the only difference is color and texture, because an atlas can be used, and color can be in vertex colors (the mesher populates them using the `color` property of `Voxel`).

**Comment by JoseETeixeira** [comment]
Is there a way to add new voxels at runtime to the VoxelLibrary ? Create voxel and bake don't appear to do it. 
The idea is : 
- Read data from json (color and position)
- Create a new voxel at the library 
- Read the library and set the voxel to the created one
I added this as a script in my VoxelLibrary


```
extends VoxelLibrary
func _init():
	var file = File.new()
	file.open("res://blocky_game/blocks/ancient_temple/ancient_temple.voxenko", file.READ)
	var json = file.get_as_text()
	var json_result = JSON.parse(json).result
	file.close()
	for voxel in json_result:
		voxel_count = voxel_count + 1;
		create_voxel(int(voxel.colour),String(voxel.colour))
	bake()

```

**Comment by Zylann** [comment]
> Is there a way to add new voxels at runtime to the VoxelLibrary ?

It is possible to do it before the terrain starts to use it, but not afterward (or requires the entire terrain to be reloaded and remeshed). You may also do this *after* the existing data was loaded. `_init` runs before the resource is loaded so items added from here will be deleted.

So the sequence should be:
1) Load VoxelLibrary (it will contain base voxels you may have configured in editor)
2) Add extra voxels from code like you did
3) Instantiate the scene that contains the terrain
4) Assign the VoxelLibrary to it
5) Add the scene to the tree

> then I would set the colors manually

Well you have to decide... why would you not be able to also set colors automatically?

Also this issue is getting full of questions not even related to the title, perhaps you could open more specific ones or jump on Discord

**Comment by JoseETeixeira** [comment]
Thanks a lot, I didn't know you had a discord server, can you link it? I got a ton of questions as I'm just getting started on voxels

**Comment by Zylann** [comment]
It's on the readme: https://github.com/Zylann/godot_voxel#readme
I might not be able to answer instantly tho
